File: scripts/scrape_themes.py
import json
from collections.abc import Callable
from dataclasses import dataclass
from urllib.parse import urljoin, urlparse

from scrapling.fetchers import StealthyFetcher
from scrapling.parser import Selector


@dataclass(frozen=True)
class SourceConfig:
    url: str
    article_selector: str
    link_selector: str
    headline_selector: str
    summary_selector: str
    image_selector: str


SOURCES: tuple[SourceConfig, ...] = (
    SourceConfig("https://techcrunch.com/latest/", "article", "a", "h2, h3", "p", "img"),
    SourceConfig("https://www.theverge.com/tech", "article", "a", "h2, h3", "p", "img"),
    SourceConfig("https://arstechnica.com/gadgets/", "article", "a", "h2, h3", "p", "img"),
)


def normalize_url(base_url: str, value: str) -> str:
    cleaned = value.strip()
    if not cleaned:
        return ""
    candidate = urljoin(base_url, cleaned)
    return candidate if urlparse(candidate).scheme in {"http", "https"} else ""


def first_text(node: Selector, selector: str) -> str:
    match = node.css(selector).first
    return " ".join(match.text.split()) if match is not None else ""


def image_urls(node: Selector, source_url: str, selector: str) -> list[str]:
    values: list[str] = []
    for image in node.css(selector):
        raw = image.attrib.get("src") or image.attrib.get("data-src") or ""
        normalized = normalize_url(source_url, raw)
        if normalized and normalized not in values:
            values = [*values, normalized]
        if len(values) == 2:
            break
    return values


def parse_source(html: str, source: SourceConfig) -> list[dict[str, object]]:
    page = Selector(html)
    candidates: list[dict[str, object]] = []
    for article in page.css(source.article_selector):
        link = article.css(source.link_selector).first
        source_url = normalize_url(source.url, link.attrib.get("href", "") if link else "")
        headline = first_text(article, source.headline_selector)
        if not source_url or not headline:
            continue
        candidates = [*candidates, {
            "sourceUrl": source_url,
            "headline": headline,
            "summary": first_text(article, source.summary_selector),
            "referenceImageUrls": image_urls(article, source.url, source.image_selector),
        }]
    return candidates


def fetch_html(url: str) -> str:
    page = StealthyFetcher.fetch(url, headless=True, network_idle=True)
    return page.html_content


def scrape_themes(fetch: Callable[[str], str]) -> list[dict[str, object]]:
    by_url: dict[str, dict[str, object]] = {}
    for source in SOURCES:
        for candidate in parse_source(fetch(source.url), source):
            by_url = {**by_url, str(candidate["sourceUrl"]): candidate}
    return list(by_url.values())


def main() -> None:
    payload = {"candidates": scrape_themes(fetch_html)}
    print(json.dumps(payload, ensure_ascii=False, separators=(",", ":")))


if __name__ == "__main__":
    main()

File: src/lib/scraping/scrapeThemes.ts
import * as cheerio from 'cheerio';

export interface SourceConfig {
  url: string;
  articleSelector: string;
  linkSelector: string;
  headlineSelector: string;
  summarySelector: string;
  imageSelector: string;
}

export interface ScrapedCandidate {
  sourceUrl: string;
  headline: string;
  summary: string;
  referenceImageUrls: string[];
}

const MAX_REFERENCE_IMAGES = 2;

export const SOURCES: readonly SourceConfig[] = [
  {
    url: 'https://techcrunch.com/latest/',
    articleSelector: 'article',
    linkSelector: 'a',
    headlineSelector: 'h2, h3',
    summarySelector: 'p',
    imageSelector: 'img',
  },
  {
    url: 'https://www.theverge.com/tech',
    articleSelector: 'article',
    linkSelector: 'a',
    headlineSelector: 'h2, h3',
    summarySelector: 'p',
    imageSelector: 'img',
  },
  {
    url: 'https://arstechnica.com/gadgets/',
    articleSelector: 'article',
    linkSelector: 'a',
    headlineSelector: 'h2, h3',
    summarySelector: 'p',
    imageSelector: 'img',
  },
];

export function normalizeUrl(baseUrl: string, value: string): string {
  const trimmed = value.trim();
  if (!trimmed) {
    return '';
  }
  try {
    const candidate = new URL(trimmed, baseUrl);
    return candidate.protocol === 'http:' || candidate.protocol === 'https:'
      ? candidate.toString()
      : '';
  } catch {
    return '';
  }
}

function firstText(article: cheerio.Cheerio<import('domhandler').Element>, selector: string): string {
  return article.find(selector).first().text().trim().replace(/\s+/g, ' ');
}

function imageUrls(
  $: cheerio.CheerioAPI,
  article: cheerio.Cheerio<import('domhandler').Element>,
  sourceUrl: string,
  selector: string,
): string[] {
  const normalized = article
    .find(selector)
    .toArray()
    .map((element) => $(element).attr('src') ?? $(element).attr('data-src') ?? '')
    .map((raw) => normalizeUrl(sourceUrl, raw))
    .filter((url) => url.length > 0);

  return Array.from(new Set(normalized)).slice(0, MAX_REFERENCE_IMAGES);
}

export function parseSource(html: string, source: SourceConfig): ScrapedCandidate[] {
  const $ = cheerio.load(html);

  return $(source.articleSelector)
    .toArray()
    .map((element) => {
      const article = $(element) as cheerio.Cheerio<import('domhandler').Element>;
      const href = article.find(source.linkSelector).first().attr('href') ?? '';
      return {
        sourceUrl: normalizeUrl(source.url, href),
        headline: firstText(article, source.headlineSelector),
        summary: firstText(article, source.summarySelector),
        referenceImageUrls: imageUrls($, article, source.url, source.imageSelector),
      };
    })
    .filter((candidate) => candidate.sourceUrl.length > 0 && candidate.headline.length > 0);
}

async function defaultFetchHtml(url: string): Promise<string> {
  const response = await fetch(url);
  if (!response.ok) {
    throw new Error(`scrapeThemes: request to ${url} failed with status ${response.status}`);
  }
  return response.text();
}

export async function scrapeThemes(
  fetchHtml: (url: string) => Promise<string> = defaultFetchHtml,
  sources: readonly SourceConfig[] = SOURCES,
): Promise<ScrapedCandidate[]> {
  const bySourceUrl = new Map<string, ScrapedCandidate>();

  for (const source of sources) {
    const html = await fetchHtml(source.url);
    for (const candidate of parseSource(html, source)) {
      bySourceUrl.set(candidate.sourceUrl, candidate);
    }
  }

  return Array.from(bySourceUrl.values());
}
